src/sundials_matrix_wrapper.cpp

```diff
@@ -419,81 +399,154 @@ void SUNMatrixWrapper::sparse_multiply(SUNMatrixWrapper *C,
+    sunindextype n = B.columns();
     
-    auto w = std::vector<sunindextype>(m);
-    auto x = std::vector<realtype>(m);
+    auto w = std::vector<sunindextype>(m); // sparsity of C(:,j)
+    auto x = std::vector<realtype>(m); // entries in C(:,j)
     
     for (j = 0; j < n; j++)
     {
         Cp[j] = nnz;                          /* column j of C starts here */
-        if ((Bp[j+1] > Bp[j]) && (nnz + m > C->capacity()))
+        if ((Bp[j+1] > Bp[j]) && (nnz + m > C.capacity()))
         {
             /*
              * if memory usage becomes a concern, remove the factor two here,
              * as it effectively trades memory efficiency against less
              * reallocations
              */
-            C->reallocate(2*C->capacity() + m);
+            C.reallocate(2*C.capacity() + m);
             // all pointers will change after reallocation
-            Cx = C->data();
-            Ci = C->indexvals();
-            Cp = C->indexptrs();
+            Cx = C.data();
+            Ci = C.indexvals();
+            Cp = C.indexptrs();
         }
         for (p = Bp[j]; p < Bp[j+1]; p++)
         {
-            nnz = scatter(Bi[p], Bx[p], w.data(), x.data(), j+1, C, nnz);
+            nnz = scatter(Bi[p], Bx[p], w.data(), gsl::make_span(x), j+1, &C,
+                          nnz);
             assert(nnz - Cp[j] <= m);
         }
         for (p = Cp[j]; p < nnz; p++)
             Cx[p] = x[Ci[p]]; // copy data to C
     }
     Cp[n] = nnz;
+    assert(nnz <= C.capacity());
     /*
      * do not reallocate here since we rather keep a matrix that is a bit
      * bigger than repeatedly resizing this matrix.
      */
 }
 
+void SUNMatrixWrapper::sparse_add(const SUNMatrixWrapper &A, realtype alpha,
+                                  const SUNMatrixWrapper &B, realtype beta) {
+    // matrix_ == nullptr is allowed on the first call
+    if (!A.matrix_ || !B.matrix_)
+        return;
+
+    sunindextype nrows = rows();
+    sunindextype ncols = columns();
+
+    check_csc(this, "sparse_multiply", "C");
+    check_csc(&A, "sparse_multiply", "A");
+    check_csc(&B, "sparse_multiply", "B");
+
+    check_dim(nrows, A.rows(), "rows", "rows", "C", "A");
+    check_dim(nrows, B.rows(), "rows", "rows", "C", "B");
+    check_dim(ncols, A.columns(), "columns", "columns", "C", "A");
+    check_dim(ncols, B.columns(), "columns", "columns", "C", "B");
+    
+    if (ncols == 0 || nrows == 0 ||
+        (A.num_nonzeros() + B.num_nonzeros() == 0))
+        return; // nothing to do
+    
+
+    /* see https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/CSparse/Source/cs_add.c
+     * modified such that we don't need to use CSparse memory structure and can
+     * work with preallocated C. This should minimize number of necessary
+     * reallocations as we can assume that C doesn't change size.
+     */
+    
+    sunindextype nnz = 0; // this keeps track of the nonzero index in C
+    
+    sunindextype j;
```

Comment:
Not urgent to be addressed, but general remark:
I personally prefer use looping variables such as "iCol" or "iData", making it clear right away what is looped over, rather than "i", "j", "k". Would be happy, if we could adhere to something like that when writing new code. :)

Reply:
I generally agree with you. Most of this code is copy pasted though so the tradeoff is between being able to compare the implemenation to the original code and following this guidelines and I opted with having the code more comparable to the original implementation.